server/index.js

```diff
@@ -84,24 +88,104 @@ app.get('/test', (req, res) => {
   res.redirect('/api')
 })
 
+
+// In frontend, location may be a string/ empty string, or a LocationObject containing fields including
+// twitterString : a string/ empty string
+// (test for location known with: if (location || location.twitterString) ... )
+// Better is to test for location specificity
+// location.specficity = 0 => no location
+// location.specficity = 10 => full postcode or lat/long
+// But test, eg: if (!location || location < x) ... since string.prop = undefined and (undefined < x) == false
 app.get('/api', async (req, res) => {
   const { userData, oauthAccessToken, oauthAccessTokenSecret } = req.session
   if (!userData) {
+    console.log(403);
     return res.status(403).send('You are not logged in with Twitter')
   }
   try {
     const followerIds = await twitter.getFollowerIds(userData.screen_name, oauthAccessToken, oauthAccessTokenSecret)
     const matchedIds = await matcher(followerIds)
-    res.send({ total: followerIds.length, matched: matchedIds.length })
+console.log('>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> query:',req.query);
+    let location = [
+      req.query,                         // query: We assume any query passed will include one+ of pc, p7, pc8, latlong
+      req.session.location,              // session: NB Any location object from session may be partial
+      userData.location,                 // userData: is the twitterString, ie location as it appears on user's Twitter profile
+      LocationObject()
```

Comment:
I don't think you need to pass `LocationObject()` in here. I could be wrong. but i think `populateLocationObject` could (and should) work by creating a fresh `LocationObject` from the array that's passed in

Reply:
unless i'm mistaken, this is going to return the "most specific" out of `req.query`, `req.session.location`, and `userData.location`. Is that right? I would expect `query` to take precedence, if it's set. Also, if we have already set `req.session.location`, is it worth doing another lookup with it?

Reply:
I would propose we:

1. Geocode `req.query`, if set
2. Otherwise geocode `userData.location`
3. Save this location in the session, and use it in other parts of the app. Don't re-geocode data in the session.